Stratis.Bitcoin/Features/Consensus/CoinViews/CachedCoinView.cs

```diff
@@ -8,269 +8,355 @@
-			public TxOut[] OriginalOutputs;
-		}
-
-		private readonly Stratis.Bitcoin.Utilities.ReaderWriterLock lockobj;
-		private readonly Dictionary<uint256, CacheItem> unspents;
-		private uint256 blockHash;
-		private uint256 innerBlockHash;
-		private readonly CoinView inner;
-		private readonly StakeChainStore stakeChainStore;
-
-		public CachedCoinView(DBreezeCoinView inner, StakeChainStore stakeChainStore = null)
-		{
-			Guard.NotNull(inner, nameof(inner));
-
-			this.inner = inner;
-			this.stakeChainStore = stakeChainStore;
-			this.MaxItems = 100000;
-			this.lockobj = new Stratis.Bitcoin.Utilities.ReaderWriterLock();
-			this.unspents = new Dictionary<uint256, CacheItem>();
-			this.PerformanceCounter =  new CachePerformanceCounter();
-		}
-
-		/// <summary>
-		/// This is used for testing the coin view
-		/// it allows a coinview that only has in memory entries
-		/// </summary>
-		public CachedCoinView(InMemoryCoinView inner, StakeChainStore stakeChainStore = null)
-		{
-			Guard.NotNull(inner, nameof(inner));
-
-			this.inner = inner;
-			this.stakeChainStore = stakeChainStore;
-			this.MaxItems = 100000;
-			this.lockobj = new Stratis.Bitcoin.Utilities.ReaderWriterLock();
-			this.unspents = new Dictionary<uint256, CacheItem>();
-			this.PerformanceCounter = new CachePerformanceCounter();
-		}
-
-		public CoinView Inner => this.inner;
-
-		public override async Task<FetchCoinsResponse> FetchCoinsAsync(uint256[] txIds)
-		{
-			Guard.NotNull(txIds, nameof(txIds));
-			
-			FetchCoinsResponse result = null;
-			UnspentOutputs[] outputs = new UnspentOutputs[txIds.Length];
-			List<int> miss = new List<int>();
-			List<uint256> missedTxIds = new List<uint256>();
-			using(this.lockobj.LockRead())
-			{
-				this.WaitOngoingTasks();
-				for(int i = 0; i < txIds.Length; i++)
-				{
-					CacheItem cache;
-					if(!this.unspents.TryGetValue(txIds[i], out cache))
-					{
-						miss.Add(i);
-						missedTxIds.Add(txIds[i]);
-					}
-					else
-					{
-						outputs[i] = cache.UnspentOutputs == null ? null :
-									 cache.UnspentOutputs.IsPrunable ? null :
-									 cache.UnspentOutputs.Clone();
-					}
-				}
-				this.PerformanceCounter.AddMissCount(miss.Count);
+    /// <summary>
+    /// Cache layer for coinview prevents too frequent updates of the data in the underlaying storage.
+    /// </summary>
+    public class CachedCoinView : CoinView, IBackedCoinView
+    {
+        /// <summary>
+        /// Item of the coinview cache that holds information about the unspent outputs 
+        /// as well as the status of the item in relation to the underlaying storage.
+        /// </summary>
+        private class CacheItem
+        {
+            /// <summary>Information about transaction's outputs. Spent outputs are nulled.</summary>
+            public UnspentOutputs UnspentOutputs;
+
+            /// <summary><c>true</c> if the unspent output information is stored in the underlaying storage, <c>false</c> otherwise.</summary>
```

Comment:
I would not necessarily say "underlaying storage" here, though practically it is storage but the cash coin view does not know the underline view is storage.

Reply:
I know, it was intentional, and you don't like it, I change it :)

Reply:
ok, won't, you merged it already